storage/storage.c

```diff
@@ -714,24 +818,65 @@ static secbool __wur derive_kek_unlock(
     const uint8_t *pin, size_t pin_len, const uint8_t *storage_salt,
     const uint8_t *ext_salt, uint8_t stretched_pin[SHA256_DIGEST_LENGTH],
     secbool privileged_bhk) {
-  stretch_pin(pin, pin_len, storage_salt, ext_salt, stretched_pin,
-              privileged_bhk);
+  secbool ret = secfalse;
+  mcu_stretch_pin(pin, pin_len, storage_salt, ext_salt, stretched_pin,
+                  privileged_bhk);
+#if USE_OPTIGA || USE_TROPIC
+  uint32_t pin_index = 0;
+#if STRETCHED_PIN_COUNT > 1
+  uint32_t pin_fails = 0;
+  if (sectrue != pin_get_fails(&pin_fails)) {
+    goto cleanup;
+  }
+  pin_index = pin_fails - 1;  // The counter has already been incremented
+#endif
+#endif
 #if USE_OPTIGA
-  optiga_pin_result ret = optiga_pin_verify(ui_progress, stretched_pin);
-  if (ret != OPTIGA_PIN_SUCCESS) {
+  optiga_pin_result optiga_ret = optiga_stretch_pin(ui_progress, stretched_pin);
+  if (optiga_ret != OPTIGA_PIN_SUCCESS) {
     memzero(stretched_pin, SHA256_DIGEST_LENGTH);
 
-    if (ret == OPTIGA_PIN_COUNTER_EXCEEDED) {
+    if (optiga_ret == OPTIGA_PIN_COUNTER_EXCEEDED) {
       // Unreachable code. Wipe should have already been triggered in unlock().
       storage_wipe();
       show_pin_too_many_screen();
     }
-    ensure(ret == OPTIGA_PIN_INVALID ? sectrue : secfalse,
+    ensure(optiga_ret == OPTIGA_PIN_INVALID ? sectrue : secfalse,
+           "optiga_stretch_pin failed");
+    goto cleanup;
+  }
+#endif
+#if USE_TROPIC
+  if (!tropic_stretch_pin(ui_progress, pin_index, stretched_pin)) {
+    goto cleanup;
+  }
+#endif
+#if USE_OPTIGA
+  optiga_ret = optiga_pin_verify(ui_progress, pin_index, stretched_pin);
+  if (optiga_ret != OPTIGA_PIN_SUCCESS) {
+    memzero(stretched_pin, SHA256_DIGEST_LENGTH);
```

Comment:
ISTM that this should go into the `cleanup:` section. Otherwise the `stretched_pin` doesn't get cleaned up in case of later failures. Also same memzero on line 837 doesn't need to be there.

Reply:
This part of code was significantly changed in https://github.com/trezor/trezor-firmware/commit/04a818b279c3ac4fc42bf01e2fe1c9283a280ded and I believe it has fixed the issue.